src/main/java/org/elasticsearch/common/compress/CompressedString.java

```diff
@@ -120,14 +102,19 @@ public boolean equals(Object o) {
 
         CompressedString that = (CompressedString) o;
 
-        if (!Arrays.equals(bytes, that.bytes)) return false;
+        if (Arrays.equals(compressed(), that.compressed())) {
+            return true;
+        }
 
-        return true;
+        return Arrays.equals(uncompressed(), that.uncompressed());
     }
 
     @Override
     public int hashCode() {
-        return bytes != null ? Arrays.hashCode(bytes) : 0;
+        if (hashCode == 0) {
```

Comment:
Isn't it possible to have a hashcode of 0? In that corner case we would just do more work (decompress on every hashCode() call)?